src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/HandlerOrchestrator.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClientX.Handlers.TransportCreation;
+
+namespace Microsoft.Data.SqlClientX.Handlers
+{
+    internal class HandlerOrchestrator
+    {
+        /// <summary>
+        /// Class responsible to create the chain of handlers, initiate chain, retain handler-context, maintain 
+        /// routing count and history of all handler contexts for all handlers in the chain
+        /// </summary>
+        private  IHandler<HandlerRequest> _initialHandler;
+        private ConnectionHandlerContext _initialConnectionHandlerContext;
+        private InternalConnectionContext _internalConnectionContext;
+        private int _routingCount;
+        private List<ConnectionHandlerContext> _routingHistory;
+
+        public HandlerOrchestrator(IHandler<HandlerRequest> initialHandler, ConnectionHandlerContext initialConnectionHandlerContext, InternalConnectionContext internalConnectionContext, List<ConnectionHandlerContext> history)
+        {
+            _initialHandler = initialHandler;
+            _initialConnectionHandlerContext = initialConnectionHandlerContext;
+            _internalConnectionContext = internalConnectionContext;
+            _routingHistory = history;
+            _routingCount = 0;
+        }
+
+        /// <summary>
+        /// Intiates the pre-defined chain of handlers
+        /// </summary>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        /// <exception cref="HandlerNotFoundException"></exception>
+        public async Task ProcessRequestAsync(bool isAsync, CancellationToken ct) 
+        {
+            CreateHandlerChain();
+            if (_initialHandler != null)
+            {
+                await _initialHandler.Handle(_initialConnectionHandlerContext, isAsync, ct).ConfigureAwait(false);
+            }
+            else
+            {
+                //TODO: Deal with missing handler exception in the connector and populate relevant SqlException in future.
+                throw new HandlerNotFoundException();
+            }
+        }
+
+        /// <summary>
+        /// Creates the chain of handlers in the order definition
+        /// </summary>
+        private void CreateHandlerChain()
+        {
+            // List of handlers to be added to the chain, ex : DSH -> TCH -> Prelogin -> Login -> null
+            List<IHandler<ConnectionHandlerContext>> handlers = new List<IHandler<ConnectionHandlerContext>>
```

Comment:
I'm gonna be honest here, this seems vastly overcomplicated for what its doing. Rather than create a list, iterate over it, etc, etc ... just create them and wire them up as you go.

```
_initialHandler = new DataSourceParsingHandler();
var  transportHandler = new TransportCreationHandler();
_initialHandler.NextHandler = transportHandler;
var preLoginHandler = new PreLoginHandler();
_transportHandler.Next = preLoginHandler;
```

or something like that

Reply:
Agreed on the overcomplication. My idea was that there will be multiple handlers in the chain DSPHandler->SSRP->TransPort->PreLogin->Login.. etc. so 10 statements just creating the chain but as a list we can have one single point of definition.

Reply:
I think this is a situation where readability wins out over rule-of-3. If the chain were to be dynamic, then I would agree that this approach is better. But since it's static, I think it makes more sense to just show x connects to y connects to z.

It's ok if you want to take this as part of the next PR that makes chain creation happen once.

Reply:
Will make the change.